File: android/src/main/java/app/capgo/proximity/CapacitorProximityPlugin.java
package app.capgo.proximity;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "CapacitorProximity")
public class CapacitorProximityPlugin extends Plugin {

    private CapacitorProximity implementation;

    @Override
    public void load() {
        super.load();
        implementation = new CapacitorProximity(getActivity());
    }

    @Override
    protected void handleOnDestroy() {
        if (implementation != null) {
            implementation.stopMonitoring();
        }
        super.handleOnDestroy();
    }

    @PluginMethod
    public void enable(PluginCall call) {
        bridge.executeOnMainThread(() -> implementation.enable(call));
    }

    @PluginMethod
    public void disable(PluginCall call) {
        bridge.executeOnMainThread(() -> implementation.disable(call));
    }

    @PluginMethod
    public void getStatus(PluginCall call) {
        call.resolve(implementation.getStatus());
    }

    @PluginMethod
    public void getPluginVersion(PluginCall call) {
        JSObject result = new JSObject();
        result.put("version", implementation.getPluginVersion());
        call.resolve(result);
    }
}

File: ios/Sources/CapacitorProximityPlugin/CapacitorProximityPlugin.swift
import Capacitor
import Foundation

@objc(CapacitorProximityPlugin)
public class CapacitorProximityPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorProximityPlugin"
    public let jsName = "CapacitorProximity"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "enable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = CapacitorProximity()

    deinit {
        implementation.disable()
    }

    @objc func enable(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            if self.implementation.enable() {
                call.resolve()
                return
            }

            call.reject("Proximity sensor not available on this device.")
        }
    }

    @objc func disable(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.implementation.disable()
            call.resolve()
        }
    }

    @objc func getStatus(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve([
                "available": self.implementation.isAvailable(),
                "enabled": self.implementation.isEnabled(),
                "platform": "ios"
            ])
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": implementation.getPluginVersion()
        ])
    }
}
